gym/envs/phys2d/bipedal_walker.py

```diff
@@ -0,0 +1,731 @@
+        p2 = self.terrain_points[1:]
+        v1 = hull_pos - p1
+        v2 = p2 - p1
+        v3 = jp.array([-directions.T[1], directions.T[0]]).T
+        distances = jp.cross(v2, v1)[:, None] / (v2 @ v3.T)
+        intersect_points = hull_pos + np.einsum("ij,ki->ikj", directions, distances)
+        valid_intersects = jp.array(
+            [
+                (
+                    (self.terrain_points[:-1, 0] <= intersect_points[i][:, 0])
+                    & (intersect_points[i][:, 0] <= self.terrain_points[1:, 0])
+                )
+                for i in range(len(directions))
+            ]
+        ).T
+        min_distances = jp.minimum(
+            LIDAR_RANGE, jp.where(valid_intersects, distances, LIDAR_RANGE)
+        ).min(axis=0)
+        normed_min_distances = min_distances / LIDAR_RANGE
+        return normed_min_distances
+
+    def _get_state(self, info):
+        left_leg_ground_contact, right_leg_ground_contact = (
+            info.contact.vel[(LEFT_LEG, RIGHT_LEG), 2] != 0
+        ).astype(int)
+        joint_angle, joint_vel = self.sys.joints[0].angle_vel(self.qp)
+        joint_angle += [0, 1, 0, 1]
+        joint_vel /= jp.array([6, 4, 6, 4])
+
+        state = [
+            self._get_angle(self.qp.rot[HULL]),
+            self.qp.ang[HULL][1] / 200,
+            *self._to_2d(0.3 * self.qp.vel[HULL]),
+            joint_angle[LEFT_HIP],
+            joint_vel[LEFT_HIP],
+            joint_angle[LEFT_KNEE],
+            joint_vel[LEFT_KNEE],
+            left_leg_ground_contact,
+            joint_angle[RIGHT_HIP],
+            joint_vel[RIGHT_HIP],
+            joint_angle[RIGHT_KNEE],
+            joint_vel[RIGHT_KNEE],
+            right_leg_ground_contact,
+            *self._get_lidar(),
+        ]
+        assert len(state) == 24
+        return state
+
+    def _generate_terrain(self, hardcore):
+        self.terrain_x = np.linspace(
+            -INITIAL_X, TERRAIN_SIZE - INITIAL_X, TERRAIN_LENGTH
+        )
+        velocity = 0
+        y = 0
+        self.terrain_y = [0] * TERRAIN_STARTPAD
+
+        for _ in range(TERRAIN_LENGTH - TERRAIN_STARTPAD):
+            velocity += (
+                -0.2 * velocity
+                - 0.01 * np.sign(y)
+                + self.np_random.uniform(-1, 1) / TERRAIN_SCALE
+            )
+            y += velocity
+            self.terrain_y.append(y)
+
+        self.terrain_y = jp.array(self.terrain_y)
+        self.terrain_points = jp.array([self.terrain_x, self.terrain_y]).T
+        terrain_map = jp.repeat(self.terrain_y, TERRAIN_LENGTH).tolist()
+        return terrain_map
+
+    def reset(
+        self,
+        *,
+        seed: Optional[int] = None,
+        return_info: bool = False,
+        options: Optional[dict] = None,
+    ):
+        super().reset(seed=seed)
+        self.renderer.reset()
+        self.prev_shaping = None
+
+        terrain_map = self._generate_terrain(self.hardcore)
+
+        _SYSTEM_CONFIG = """
```

Comment:
This should be moved to `init` as we don't understand users change the parameters between resets